rest/changes_api_test.go

```diff
@@ -556,6 +546,158 @@ func postChangesAdminChannelGrant(t *testing.T, rt *RestTester) {
+	_ = rt.putDoc("mix-2", fmt.Sprintf(`{"_rev":%q, "channel":["HBO"]}`, mix2.Rev))
+	_ = rt.putDoc("mix-4", fmt.Sprintf(`{"_rev":%q}`, mix4.Rev))
+
+	// Validate that tombstones are also not sent as part of backfill:
+	//  Case 1: Delete a document in a single channel (e.g. pbs-3), and validate it doesn't get included in backfill
+	//  Case 2: Delete a document in a multiple channels (e.g. mix-3), and validate it doesn't get included in backfill
+	rt.deleteDoc(pbs3.ID, pbs3.Rev)
+	rt.deleteDoc(mix3.ID, mix3.Rev)
+
+	// Test Scenario:
+	//   1. Document mix-5 is in channels PBS, HBO (rev 1)
+	//   2. Document is deleted (rev 2)
+	//   3. Document is recreated, only in channel PBS (rev 3)
+	mix5 := rt.putDoc("mix-5", `{"channel":["PBS","HBO"]}`)
+	rt.deleteDoc(mix5.ID, mix5.Rev)
+	_ = rt.putDoc(mix5.ID, `{"channel":["PBS"]}`)
+
+	// Test Scenario:
+	//   1. Document mix-6 is in channels PBS, HBO (rev 1)
+	//   2. Document is updated only be in channel HBO (rev 2)
+	//   3. Document is updated AGAIN to remove all channels (rev 3)
+	mix6 := rt.putDoc("mix-6", `{"channel":["PBS","HBO"]}`)
+	mix6 = rt.putDoc(mix6.ID, fmt.Sprintf(`{"_rev":%q, "channel":["HBO"]}`, mix6.Rev))
+	_ = rt.putDoc(mix6.ID, fmt.Sprintf(`{"_rev":%q}`, mix6.Rev))
+
+	// Test Scenario:
+	//   1. Delete abc-2 from channel ABC
+	//   2. Update abc-3 to remove from channel ABC
+	rt.deleteDoc(abc2.ID, abc2.Rev)
+	_ = rt.putDoc(abc3.ID, fmt.Sprintf(`{"_rev":%q}`, abc3.Rev))
+
+	// Issue changes request and check the results
+	expectedResults := []string{
+		`{"seq":11,"id":"abc-1","changes":[{"rev":"1-0143105976caafbda3b90cf82948dc64"}]}`,
+		`{"seq":26,"id":"abc-2","deleted":true,"removed":["ABC"],"changes":[{"rev":"2-6055be21d970eb690f48452505ea02ed"}]}`,
+		`{"seq":27,"id":"abc-3","removed":["ABC"],"changes":[{"rev":"2-09b89154aa9a0e1620da0d86528d406a"}]}`,
+	}
+	changes, err := rt.WaitForChanges(len(expectedResults), "/db/_changes", "bernard", false)
+	require.NoError(t, err, "Error retrieving changes results")
+
+	for index, result := range changes.Results {
+		var expectedChange db.ChangeEntry
+		assert.NoError(t, base.JSONUnmarshal([]byte(expectedResults[index]), &expectedChange))
+		assert.Equal(t, expectedChange.Seq, result.Seq)
+		assert.Equal(t, expectedChange.ID, result.ID)
+		assert.Equal(t, expectedChange.Changes, result.Changes)
+		assert.Equal(t, expectedChange.Deleted, result.Deleted)
+		assert.Equal(t, expectedChange.Removed, result.Removed)
+	}
+
+	// Update the user doc to grant access to PBS, HBO in addition to ABC
+	response := rt.SendAdminRequest(http.MethodPut, "/db/_user/bernard", `{"admin_channels":["ABC", "PBS", "HBO"]}`)
+	assertStatus(t, response, http.StatusOK)
+
+	// Issue a new changes request with since=last_seq ensure that user receives all records for channels PBS, HBO.
+	expectedResults = []string{
+		`{"seq":"28:1","id":"pbs-1","changes":[{"rev":"1-82214a562e80c8fa7b2361719847bc73"}]}`,
```

Comment:
Asserting on specific sequence values isn't reliable when using the default (batching) sequence allocator - depending on test timing, there may be unused sequences, which would throw off your verification here.  

You can avoid this by disabling sequence batching for this test, in the same way done here: https://github.com/couchbase/sync_gateway/blob/d0d138c96bd2dc1ef8a2d1090c485d2ccb914708/rest/replication_api_test.go#L543-L546

Reply:
Makes sense, included a function to suspend sequence batching for multi-RT tests